File: apps/api/src/recomendacoes/ia/justificativa.ts
import type { CategoriaProduto } from '@prisma/client';
import type { Contexto, Fatores, ProdutoIA, Scored } from './types';

const CATEGORIA_LABEL: Record<CategoriaProduto, string> = {
  RENDA_FIXA: 'renda fixa',
  RENDA_VARIAVEL: 'renda variável',
  FUNDOS: 'fundos de investimento',
  PREVIDENCIA: 'previdência privada',
  ESTRUTURADOS: 'produtos estruturados',
  CAMBIO: 'câmbio',
};

const brl = new Intl.NumberFormat('pt-BR', {
  style: 'currency',
  currency: 'BRL',
  maximumFractionDigits: 0,
});

/** "R$ 14,5 milhões" / "R$ 800 mil" / "R$ 350" — escala humana */
function humanizarPatrimonio(valor: number): string {
  if (valor >= 1_000_000_000) {
    const b = valor / 1_000_000_000;
    return `R$ ${b >= 10 ? Math.round(b) : b.toFixed(1).replace('.', ',')} bilhões`;
  }
  if (valor >= 1_000_000) {
    const m = valor / 1_000_000;
    return `R$ ${m >= 10 ? Math.round(m) : m.toFixed(1).replace('.', ',')} milhões`;
  }
  if (valor >= 100_000) {
    return `R$ ${Math.round(valor / 1_000)} mil`;
  }
  return brl.format(valor);
}

function pctCategoria(ctx: Contexto, categoria: CategoriaProduto): number {
  if (ctx.cliente.patrimonio === 0) return 0;
  const total = ctx.posicoes
    .filter((p) => p.categoria === categoria)
    .reduce((acc, p) => acc + p.valor, 0);
  return Math.round((total / ctx.cliente.patrimonio) * 100);
}

// ============================================================
// Frases técnicas (uma por fator) — usadas no "Como cheguei nisso"
// ============================================================

export function fraseFator(
  fator: keyof Fatores,
  valor: number,
  ctx: Contexto,
  p: ProdutoIA,
): string {
  switch (fator) {
    case 'profileMatch': {
      if (valor === 1)
        return `Produto enquadrado exatamente no perfil ${ctx.cliente.perfil.toLowerCase()}.`;
      return `Produto adequado ao perfil ${ctx.cliente.perfil.toLowerCase()} (perfil mínimo: ${p.perfilMinimo.toLowerCase()}).`;
    }
    case 'diversification': {
      const pct = pctCategoria(ctx, p.categoria);
      if (valor > 0.6) {
        return `Categoria ${CATEGORIA_LABEL[p.categoria]} subponderada — exposição atual de apenas ${pct}% versus alvo do perfil.`;
      }
      if (valor > 0.3) {
        return `Há espaço para reforçar exposição em ${CATEGORIA_LABEL[p.categoria]} (${pct}% atual).`;
      }
      return `Categoria ${CATEGORIA_LABEL[p.categoria]} já próxima do alvo (${pct}%).`;
    }
    case 'yield': {
      if (valor > 0.7)
        return `Rentabilidade de ${p.rentabilidadeAno.toFixed(1)}% a.a. está entre as melhores da categoria.`;
      if (valor > 0.4)
        return `Rentabilidade de ${p.rentabilidadeAno.toFixed(1)}% a.a. é competitiva.`;
      return `Rentabilidade de ${p.rentabilidadeAno.toFixed(1)}% a.a. abaixo da média da categoria.`;
    }
    case 'liquidity': {
      const h = ctx.suitability.horizonteAnos;
      if (valor > 0.7)
        return `Liquidez ${p.liquidez} alinhada ao horizonte de ${h} anos declarado pelo cliente.`;
      if (valor < 0.4)
        return `Liquidez ${p.liquidez} mais longa que o horizonte de ${h} anos — avaliar.`;
      return `Liquidez ${p.liquidez} aceitável para o horizonte do cliente.`;
    }
    case 'cost': {
      if (p.taxaAdmin === null) return 'Sem taxa de administração.';
      if (valor > 0.7)
        return `Taxa de administração baixa (${p.taxaAdmin.toFixed(1)}% a.a.).`;
      return `Taxa de administração de ${p.taxaAdmin.toFixed(1)}% a.a. — verificar custo-benefício.`;
    }
  }
}

// ============================================================
// Justificativa em narrativa (1ª pessoa) — usada como copy principal
// ============================================================

/** Frase de abertura — varia conforme o fator com maior peso */
function abertura(
  dom: keyof Fatores,
  p: ProdutoIA,
  ctx: Contexto,
): string {
  const primeiro = ctx.cliente.nome.split(' ')[0];
  const perfil = ctx.cliente.perfil.toLowerCase();

  switch (dom) {
    case 'diversification': {
      const pct = pctCategoria(ctx, p.categoria);
      const cat = CATEGORIA_LABEL[p.categoria];
      if (pct === 0) {
        return `Olhei a carteira de ${primeiro} e a primeira coisa que me chamou atenção é a ausência total de exposição em ${cat}.`;
      }
      return `Analisando a carteira de ${primeiro}, vi que ${cat} representa apenas ${pct}% — bem abaixo do alvo pra perfil ${perfil}.`;
    }
    case 'profileMatch':
      return `O ${p.nome} tem o desenho certo pra ${primeiro}: foi pensado pra perfil ${p.perfilMinimo.toLowerCase()}, que conversa direto com o perfil ${perfil} do cliente.`;
    case 'yield': {
      const cat = CATEGORIA_LABEL[p.categoria];
      return `Dentro de ${cat}, o ${p.nome} entrega ${p.rentabilidadeAno.toFixed(1)}% a.a. — está entre os melhores que o BTG distribui na categoria.`;
    }
    case 'liquidity':
      return `Considerando o horizonte de ${ctx.suitability.horizonteAnos} anos declarado por ${primeiro}, a liquidez ${p.liquidez} do ${p.nome} cai bem.`;
    case 'cost': {
      const custo =
        p.taxaAdmin === null
          ? 'não tem taxa de administração'
          : `tem taxa de admin de apenas ${p.taxaAdmin.toFixed(1)}% a.a.`;
      return `Um ponto forte do ${p.nome} pra ${primeiro} é o custo: ${custo}, o que preserva o yield líquido.`;
    }
  }
}

/** Frase opcional de contexto sobre o cliente (patrimônio, perfil) */
function contexto(dom: keyof Fatores, ctx: Contexto): string {
  const perfil = ctx.cliente.perfil.toLowerCase();
  const patrimonio = humanizarPatrimonio(ctx.cliente.patrimonio);

  switch (dom) {
    case 'diversification':
      return `Para um perfil ${perfil} com ${patrimonio} de patrimônio, isso é uma lacuna estratégica.`;
    case 'yield':
      return `Vale considerar pra ${ctx.cliente.nome.split(' ')[0]}, que tem ${patrimonio} em jogo.`;
    case 'profileMatch':
    case 'liquidity':
    case 'cost':
      return '';
  }
}

/** Endorsement curto de um fator quando ele aparece como secundário na proposta */
function endorsement(
  fator: keyof Fatores,
  valor: number,
  p: ProdutoIA,
  ctx: Contexto,
): string {
  switch (fator) {
    case 'profileMatch':
      if (valor >= 1) return 'está enquadrado exatamente no perfil do cliente';
      if (valor >= 0.7) return 'é compatível com o perfil declarado';
      return 'é aceito pra esse perfil';
    case 'diversification':
      if (valor >= 0.6) return 'preenche um gap importante da carteira';
      if (valor >= 0.3) return 'reforça uma categoria ainda subponderada';
      return null as never; // contribuição irrelevante — não citar
    case 'yield':
      if (valor >= 0.7)
        return `a rentabilidade de ${p.rentabilidadeAno.toFixed(1)}% a.a. está no topo da categoria`;
      if (valor >= 0.4)
        return `${p.rentabilidadeAno.toFixed(1)}% a.a. é um retorno competitivo`;
      return null as never;
    case 'liquidity':
      if (valor >= 0.7)
        return `a liquidez ${p.liquidez} cabe folgadamente no horizonte de ${ctx.suitability.horizonteAnos} anos`;
      if (valor >= 0.4) return `a liquidez ${p.liquidez} é aceitável pra esse horizonte`;
      return null as never;
    case 'cost':
      if (p.taxaAdmin === null)
        return 'não tem taxa de administração penalizando o yield';
      if (valor >= 0.7)
        return `a taxa de admin de ${p.taxaAdmin.toFixed(1)}% a.a. é baixa pra categoria`;
      return null as never;
  }
}

/** Junta endorsements em pt-BR natural ("X, Y e Z") */
function juntarPtBr(itens: string[]): string {
  if (itens.length === 0) return '';
  if (itens.length === 1) return itens[0];
  if (itens.length === 2) return `${itens[0]} e ${itens[1]}`;
  return `${itens.slice(0, -1).join(', ')} e ${itens[itens.length - 1]}`;
}

/** Proposta — apresenta o produto como solução, somando os pontos secundários */
function proposta(
  dom: keyof Fatores,
  p: ProdutoIA,
  secundarios: { fator: keyof Fatores }[],
  fatores: Fatores,
  ctx: Contexto,
): string {
  const endorsements = secundarios
    .map((s) => endorsement(s.fator, fatores[s.fator], p, ctx))
    .filter((e): e is string => typeof e === 'string' && e.length > 0);

  if (endorsements.length === 0) {
    // Caso raro: nenhum fator secundário valeu citação
    if (dom === 'diversification' || dom === 'profileMatch') return '';
    return '';
  }

  const lista = juntarPtBr(endorsements);

  if (dom === 'diversification') {
    return `O ${p.nome} resolveria isso: ${lista}.`;
  }
  if (dom === 'profileMatch') {
    return `Reforçando a escolha: ${lista}.`;
  }
  // Quando o dominante já é o "ponto forte do produto" (yield/liquidity/cost),
  // a proposta fica como "soma-se a isso..."
  return `Soma-se a isso que ${lista}.`;
}

/**
 * Monta a justificativa final em 3 partes:
 *   [Abertura focada no fator dominante]  [Contexto sobre o cliente]  [Proposta com 2 endorsements]
 */
export function buildJustificativa(s: Scored, ctx: Contexto): string {
  const sortedContribs = [...s.contribs].sort((a, b) => b.contrib - a.contrib);
  const dom = sortedContribs[0].fator;
  const secundarios = sortedContribs.slice(1, 3);

  const partes = [
    abertura(dom, s.produto, ctx),
    contexto(dom, ctx),
    proposta(dom, s.produto, secundarios, s.fatores, ctx),
  ].filter((p) => p && p.length > 0);

  return partes.join(' ');
}

File: apps/ai-engine/src/justificativa.py
"""
Construtor de justificativas em pt-BR — porte da versão TypeScript.

Cada recomendação ganha:
  - frases técnicas curtas por fator (usadas no "Como cheguei nisso" do front)
  - uma justificativa em narrativa de 1ª pessoa (lead da recomendação)
"""
from .models import Cliente, Posicao, Produto, Suitability

CATEGORIA_LABEL: dict[str, str] = {
    "RENDA_FIXA": "renda fixa",
    "RENDA_VARIAVEL": "renda variável",
    "FUNDOS": "fundos de investimento",
    "PREVIDENCIA": "previdência privada",
    "ESTRUTURADOS": "produtos estruturados",
    "CAMBIO": "câmbio",
}


def humanizar_patrimonio(v: float) -> str:
    """R$ 14,5 milhões / R$ 800 mil / R$ 350"""
    if v >= 1_000_000_000:
        b = v / 1_000_000_000
        n = f"{round(b)}" if b >= 10 else f"{b:.1f}".replace(".", ",")
        return f"R$ {n} bilhões"
    if v >= 1_000_000:
        m = v / 1_000_000
        n = f"{round(m)}" if m >= 10 else f"{m:.1f}".replace(".", ",")
        return f"R$ {n} milhões"
    if v >= 100_000:
        return f"R$ {round(v / 1_000)} mil"
    s = f"R$ {v:,.0f}".replace(",", ".")
    return s


def pct_categoria(
    cliente: Cliente, posicoes: list[Posicao], categoria: str
) -> int:
    if cliente.patrimonio == 0:
        return 0
    total = sum(p.valor for p in posicoes if p.categoria == categoria)
    return round((total / cliente.patrimonio) * 100)


# ============================================================
# Frases técnicas — uma por fator (drill-down do front)
# ============================================================


def frase_fator(
    fator: str,
    valor: float,
    cliente: Cliente,
    posicoes: list[Posicao],
    p: Produto,
) -> str:
    if fator == "profileMatch":
        if valor == 1.0:
            return (
                f"Produto enquadrado exatamente no perfil "
                f"{cliente.perfil.lower()}."
            )
        return (
            f"Produto adequado ao perfil {cliente.perfil.lower()} "
            f"(perfil mínimo: {p.perfilMinimo.lower()})."
        )

    if fator == "diversification":
        pct = pct_categoria(cliente, posicoes, p.categoria)
        cat = CATEGORIA_LABEL[p.categoria]
        if valor > 0.6:
            return (
                f"Categoria {cat} subponderada — exposição atual de apenas "
                f"{pct}% versus alvo do perfil."
            )
        if valor > 0.3:
            return f"Há espaço para reforçar exposição em {cat} ({pct}% atual)."
        return f"Categoria {cat} já próxima do alvo ({pct}%)."

    if fator == "yield":
        rent = f"{p.rentabilidadeAno:.1f}".replace(".", ",")
        if valor > 0.7:
            return (
                f"Rentabilidade de {rent}% a.a. está entre as melhores da "
                f"categoria."
            )
        if valor > 0.4:
            return f"Rentabilidade de {rent}% a.a. é competitiva."
        return f"Rentabilidade de {rent}% a.a. abaixo da média da categoria."

    if fator == "liquidity":
        h = "horizonte"  # placeholder — preenchido por quem chama
        if valor > 0.7:
            return (
                f"Liquidez {p.liquidez} alinhada ao {h} declarado pelo cliente."
            )
        if valor < 0.4:
            return f"Liquidez {p.liquidez} mais longa que o {h} — avaliar."
        return f"Liquidez {p.liquidez} aceitável para o {h} do cliente."

    if fator == "cost":
        if p.taxaAdmin is None:
            return "Sem taxa de administração."
        taxa = f"{p.taxaAdmin:.1f}".replace(".", ",")
        if valor > 0.7:
            return f"Taxa de administração baixa ({taxa}% a.a.)."
        return f"Taxa de administração de {taxa}% a.a. — verificar custo-benefício."

    return ""


# ============================================================
# Justificativa em narrativa (1ª pessoa)
# ============================================================


def abertura(
    dominante: str,
    cliente: Cliente,
    posicoes: list[Posicao],
    p: Produto,
    suitability: Suitability,
) -> str:
    primeiro = cliente.nome.split(" ")[0]
    perfil = cliente.perfil.lower()
    rent = f"{p.rentabilidadeAno:.1f}".replace(".", ",")

    if dominante == "diversification":
        pct = pct_categoria(cliente, posicoes, p.categoria)
        cat = CATEGORIA_LABEL[p.categoria]
        if pct == 0:
            return (
                f"Olhei a carteira de {primeiro} e a primeira coisa que me "
                f"chamou atenção é a ausência total de exposição em {cat}."
            )
        return (
            f"Analisando a carteira de {primeiro}, vi que {cat} representa "
            f"apenas {pct}% — bem abaixo do alvo pra perfil {perfil}."
        )

    if dominante == "profileMatch":
        return (
            f"O {p.nome} tem o desenho certo pra {primeiro}: foi pensado pra "
            f"perfil {p.perfilMinimo.lower()}, que conversa direto com o "
            f"perfil {perfil} do cliente."
        )

    if dominante == "yield":
        cat = CATEGORIA_LABEL[p.categoria]
        return (
            f"Dentro de {cat}, o {p.nome} entrega {rent}% a.a. — está entre "
            f"os melhores que o BTG distribui na categoria."
        )

    if dominante == "liquidity":
        return (
            f"Considerando o horizonte de {suitability.horizonteAnos} anos "
            f"declarado por {primeiro}, a liquidez {p.liquidez} do {p.nome} "
            f"cai bem."
        )

    if dominante == "cost":
        if p.taxaAdmin is None:
            custo = "não tem taxa de administração"
        else:
            taxa = f"{p.taxaAdmin:.1f}".replace(".", ",")
            custo = f"tem taxa de admin de apenas {taxa}% a.a."
        return (
            f"Um ponto forte do {p.nome} pra {primeiro} é o custo: "
            f"{custo}, o que preserva o yield líquido."
        )

    return ""


def contexto(dominante: str, cliente: Cliente) -> str:
    perfil = cliente.perfil.lower()
    patrimonio = humanizar_patrimonio(cliente.patrimonio)
    primeiro = cliente.nome.split(" ")[0]

    if dominante == "diversification":
        return (
            f"Para um perfil {perfil} com {patrimonio} de patrimônio, "
            f"isso é uma lacuna estratégica."
        )
    if dominante == "yield":
        return f"Vale considerar pra {primeiro}, que tem {patrimonio} em jogo."
    return ""


def endorsement(
    fator: str,
    valor: float,
    p: Produto,
    suitability: Suitability,
) -> str | None:
    """Frase curta pra fator quando aparece como secundário na proposta.
    Retorna None se o fator é fraco demais pra citar."""

    if fator == "profileMatch":
        if valor >= 1.0:
            return "está enquadrado exatamente no perfil do cliente"
        if valor >= 0.7:
            return "é compatível com o perfil declarado"
        return "é aceito pra esse perfil"

    if fator == "diversification":
        if valor >= 0.6:
            return "preenche um gap importante da carteira"
        if valor >= 0.3:
            return "reforça uma categoria ainda subponderada"
        return None

    if fator == "yield":
        rent = f"{p.rentabilidadeAno:.1f}".replace(".", ",")
        if valor >= 0.7:
            return f"a rentabilidade de {rent}% a.a. está no topo da categoria"
        if valor >= 0.4:
            return f"{rent}% a.a. é um retorno competitivo"
        return None

    if fator == "liquidity":
        if valor >= 0.7:
            return (
                f"a liquidez {p.liquidez} cabe folgadamente no horizonte de "
                f"{suitability.horizonteAnos} anos"
            )
        if valor >= 0.4:
            return f"a liquidez {p.liquidez} é aceitável pra esse horizonte"
        return None

    if fator == "cost":
        if p.taxaAdmin is None:
            return "não tem taxa de administração penalizando o yield"
        if valor >= 0.7:
            taxa = f"{p.taxaAdmin:.1f}".replace(".", ",")
            return f"a taxa de admin de {taxa}% a.a. é baixa pra categoria"
        return None

    return None


def juntar_pt_br(itens: list[str]) -> str:
    if not itens:
        return ""
    if len(itens) == 1:
        return itens[0]
    if len(itens) == 2:
        return f"{itens[0]} e {itens[1]}"
    return f"{', '.join(itens[:-1])} e {itens[-1]}"


def proposta(
    dominante: str,
    p: Produto,
    secundarios: list[dict],
    fatores: dict[str, float],
    suitability: Suitability,
) -> str:
    endos: list[str] = []
    for s in secundarios:
        e = endorsement(s["fator"], fatores[s["fator"]], p, suitability)
        if e:
            endos.append(e)

    if not endos:
        return ""

    lista = juntar_pt_br(endos)

    if dominante == "diversification":
        return f"O {p.nome} resolveria isso: {lista}."
    if dominante == "profileMatch":
        return f"Reforçando a escolha: {lista}."
    return f"Soma-se a isso que {lista}."


def build_justificativa(
    scored: dict,
    cliente: Cliente,
    posicoes: list[Posicao],
    suitability: Suitability,
) -> str:
    """Monta a justificativa final em 3 partes."""
    contribs = sorted(scored["contribs"], key=lambda c: c["contrib"], reverse=True)
    dom = contribs[0]["fator"]
    secundarios = contribs[1:3]

    p: Produto = scored["produto"]
    fatores: dict[str, float] = scored["fatores"]

    partes = [
        abertura(dom, cliente, posicoes, p, suitability),
        contexto(dom, cliente),
        proposta(dom, p, secundarios, fatores, suitability),
    ]

    return " ".join(part for part in partes if part)


def preencher_frases_fator(
    scored: dict,
    cliente: Cliente,
    posicoes: list[Posicao],
    suitability: Suitability,
) -> None:
    """Substitui as frases vazias dos contribs por frases técnicas."""
    p: Produto = scored["produto"]
    fatores: dict[str, float] = scored["fatores"]

    for c in scored["contribs"]:
        frase = frase_fator(c["fator"], fatores[c["fator"]], cliente, posicoes, p)
        if c["fator"] == "liquidity":
            # frase_fator usa placeholder "horizonte" — substituir aqui
            frase = frase.replace(
                "horizonte", f"horizonte de {suitability.horizonteAnos} anos"
            )
        c["frase"] = frase
